### PR TITLE
Add function to compute the volume of a watertight mesh

### PR DESCRIPTION
**Pull request by kongsgard**
Addresses #2399 and #2063

---
This change is [Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2407)

### CODE REVIEW

File: cpp/open3d/geometry/TriangleMesh.cpp [hunk 1]
```diff
@@ -1188,6 +1188,36 @@ double TriangleMesh::GetSurfaceArea(std::vector<double> &triangle_areas) const {
...
 
+double TriangleMesh::GetVolume() const {
+    if (!IsWatertight()) {
+        utility::LogError(
+                "The mesh is not watertight, and the volume cannot be "
+                "computed.");
+    }
+
+    double volume = 0;
+    for (size_t tidx = 0; tidx < triangles_.size(); ++tidx) {
```
Comment (kongsgard) [inline cpp/open3d/geometry/TriangleMesh.cpp:1199]: This code block is easily parallelizable, and I guess OpenMP could be used here. Are the following lines all that's needed?
```
#ifdef _OPENMP
#pragma omp parallel for schedule(static)
#endif
```

**Comment by germanros1987** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- cpp/open3d/geometry/TriangleMesh.cpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/intel-isl/Open3D/pullRequest?prid=6268545&bid=20408079)

**Review by yxlao — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2407)* status: all files reviewed, 3 unresolved discussions (waiting on @kongsgard)

___
*[cpp/open3d/geometry/TriangleMesh.h, line 341 at r1](https://reviewable.io/reviews/intel-isl/open3d/2407#-MIbalIa1URMdGVhCyJs:-MIbYlGpAtnc07z2DwUB:bf4o8uc) ([raw file](https://github.com/intel-isl/open3d/blob/62b95da12842fb6e3c7630b80d40a580ee8525d5/cpp/open3d/geometry/TriangleMesh.h#L341)):*
> ```objc
>     /// checking if the origin is at the same side as the normal with respect to
>     /// the triangle.
>     double ComputeSignedVolumeOfTriangle(const Eigen::Vector3d &p0,
> ```

Can this be embedded in GetVolume?
___
*[cpp/open3d/geometry/TriangleMesh.h, line 347 at r1](https://reviewable.io/reviews/intel-isl/open3d/2407#-MIbaqNDCx7R-DPYuM5M:-MIbaqNDCx7R-DPYuM5N:brd0hcu) ([raw file](https://github.com/intel-isl/open3d/blob/62b95da12842fb6e3c7630b80d40a580ee8525d5/cpp/open3d/geometry/TriangleMesh.h#L347)):*
> ```objc
>     /// Function that computes the volume of a triangle identified by the
>     /// triangle index.
>     double GetSignedVolumeOfTriangle(size_t triangle_idx) const;
> ```

Same, can this be embedded in GetVolume?
___
*[cpp/open3d/geometry/TriangleMesh.cpp, line 1199 at r1](https://reviewable.io/reviews/intel-isl/open3d/2407#-MIZNxzeiBXMt1ozPAlN-r1-1199:-MIbZ8DK2Nf7wK8HCVLM:b-rpbszz) ([raw file](https://github.com/intel-isl/open3d/blob/62b95da12842fb6e3c7630b80d40a580ee8525d5/cpp/open3d/geometry/TriangleMesh.cpp#L1199)):*
<details><summary><i>Previously, kongsgard (Sondre Kongsgård) wrote…</i></summary><blockquote>

This code block is easily parallelizable, and I guess OpenMP could be used here. Are the following lines all that's needed?
```
#ifdef _OPENMP
#pragma omp parallel for schedule(static)
#endif
```
</blockquote></details>

Yes, that's all we need, with two catches:
- For reduction, you may need `#pragma omp parallel for reduction ...`.
- Certain MSVC compilers require signed index (e.g. `int` or `int64_t`). If `size_t` fails on our Windows CI, you may need to change to a signed integer.

File: cpp/open3d/geometry/TriangleMesh.cpp (hunk 1 above)
Comment (yxlao) [reply cpp/open3d/geometry/TriangleMesh.cpp:1211]: Yes, that's all we need, with two catches:

- For reduction, you may need `#pragma omp parallel for reduction ....`
- Certain MSVC compilers require signed indices (e.g. `int` or `int64_t`). If `size_t` fails on our Windows CI, you may need to change to signed indices.

**Review by kongsgard — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2407)* status: 2 of 4 files reviewed, 3 unresolved discussions (waiting on @yxlao)

___
*[cpp/open3d/geometry/TriangleMesh.h, line 341 at r1](https://reviewable.io/reviews/intel-isl/open3d/2407#-MIbalIa1URMdGVhCyJs:-MIcgIFu69s4WqDFxpcf:b-xr41vd) ([raw file](https://github.com/intel-isl/open3d/blob/62b95da12842fb6e3c7630b80d40a580ee8525d5/cpp/open3d/geometry/TriangleMesh.h#L341)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

Can this be embedded in GetVolume?
</blockquote></details>

Refactored to a lambda expression.
___
*[cpp/open3d/geometry/TriangleMesh.h, line 347 at r1](https://reviewable.io/reviews/intel-isl/open3d/2407#-MIbaqNDCx7R-DPYuM5M:-MIcgHhVB8qdwMLtqG29:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/62b95da12842fb6e3c7630b80d40a580ee8525d5/cpp/open3d/geometry/TriangleMesh.h#L347)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

Same, can this be embedded in GetVolume?
</blockquote></details>

Done.
___
*[cpp/open3d/geometry/TriangleMesh.cpp, line 1199 at r1](https://reviewable.io/reviews/intel-isl/open3d/2407#-MIZNxzeiBXMt1ozPAlN-r1-1199:-MIchAtu2fHtH8qlNVZU:beu5maj) ([raw file](https://github.com/intel-isl/open3d/blob/62b95da12842fb6e3c7630b80d40a580ee8525d5/cpp/open3d/geometry/TriangleMesh.cpp#L1199)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

Yes, that's all we need, with two catches:

- For reduction, you may need `#pragma omp parallel for reduction ....`
- Certain MSVC compilers require signed indices (e.g. `int` or `int64_t`). If `size_t` fails on our Windows CI, you may need to change to signed indices.
</blockquote></details>

I'll try to add this after #2408 is closed. I haven't used OpenMP before, but based on your hint I think the following should suffice:
```
double volume = 0;
size_t num_triangles = triangles_.size();
#pragma omp parallel for shared(volume, num_triangles) reduction(+ : volume)
```

**Comment by yxlao** [comment]
also adding @griegler in the loop

**Review by griegler — CHANGES_REQUESTED** [review]
Thanks @kongsgard for the PR.

> Reviewed 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2407#-:-MId-QqQ8_tCi9ia-Kll:bko56wq)* status: 3 of 4 files reviewed, 5 unresolved discussions (waiting on @kongsgard and @yxlao)

___
*[cpp/open3d/geometry/TriangleMesh.h, line 334 at r2](https://reviewable.io/reviews/intel-isl/open3d/2407#-MIczc4Z9_5cBkzweI_4:-MIczc4_9k6Qu7MIfhqd:b-ftpywr) ([raw file](https://github.com/intel-isl/open3d/blob/fb712e3e44e73b9ae8c19a6463d21fb699ad9866/cpp/open3d/geometry/TriangleMesh.h#L334)):*
> ```objc
> 
>     /// Function that computes the volume of the mesh, under the condition that
>     /// it is watertight.
> ```

Could you please add a reference to the paper/method used.
___
*[cpp/tests/geometry/TriangleMesh.cpp, line 1070 at r2](https://reviewable.io/reviews/intel-isl/open3d/2407#-MId-ER3AoUTrPm8SDjI:-MId-ER3AoUTrPm8SDjJ:bb819zt) ([raw file](https://github.com/intel-isl/open3d/blob/fb712e3e44e73b9ae8c19a6463d21fb699ad9866/cpp/tests/geometry/TriangleMesh.cpp#L1070)):*
> ```cpp
> 
> TEST(TriangleMesh, GetVolume) {
>     EXPECT_NEAR(geometry::TriangleMesh::CreateBox()->GetVolume(), 1.0, 0.01);
> ```

Could we also add a test that is a bit more complex. Eg. a sphere with sufficient subdivision should get arbitrary close to the actual volume.

**Review by kongsgard — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2407)* status: 1 of 4 files reviewed, 5 unresolved discussions (waiting on @griegler and @yxlao)

___
*[cpp/open3d/geometry/TriangleMesh.h, line 334 at r2](https://reviewable.io/reviews/intel-isl/open3d/2407#-MIczc4Z9_5cBkzweI_4:-MIdj5Tm710DaNPCE-3U:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/fb712e3e44e73b9ae8c19a6463d21fb699ad9866/cpp/open3d/geometry/TriangleMesh.h#L334)):*
<details><summary><i>Previously, griegler (Gernot) wrote…</i></summary><blockquote>

Could you please add a reference to the paper/method used.
</blockquote></details>

Done.
___
*[cpp/tests/geometry/TriangleMesh.cpp, line 1070 at r2](https://reviewable.io/reviews/intel-isl/open3d/2407#-MId-ER3AoUTrPm8SDjI:-MIdj7us7XMyG8xEU9Cl:blz3k6g) ([raw file](https://github.com/intel-isl/open3d/blob/fb712e3e44e73b9ae8c19a6463d21fb699ad9866/cpp/tests/geometry/TriangleMesh.cpp#L1070)):*
<details><summary><i>Previously, griegler (Gernot) wrote…</i></summary><blockquote>

Could we also add a test that is a bit more complex. Eg. a sphere with sufficient subdivision should get arbitrary close to the actual volume.
</blockquote></details>

Done. With the current precision the computation takes about 60ms on my computer, but I guess that's still acceptable for a unit test.

**Review by griegler — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2407)* status: 3 of 4 files reviewed, 4 unresolved discussions (waiting on @kongsgard and @yxlao)

___
*[cpp/open3d/geometry/TriangleMesh.cpp, line 1204 at r3](https://reviewable.io/reviews/intel-isl/open3d/2407#-MIdsGEE1uPgBd9MLIF9:-MIdsGEE1uPgBd9MLIFA:b699x8o) ([raw file](https://github.com/intel-isl/open3d/blob/7dfb92e37e44573df24fd5ba23fee72c72b0214c/cpp/open3d/geometry/TriangleMesh.cpp#L1204)):*
> ```cpp
>     };
> 
>     if (!IsWatertight()) {
> ```

What I get from the paper `IsWatertight` is not sufficient, I think you should also check `IsOrientable`.

**Review by kongsgard — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2407)* status: 1 of 4 files reviewed, 4 unresolved discussions (waiting on @griegler and @yxlao)

___
*[cpp/open3d/geometry/TriangleMesh.cpp, line 1199 at r1](https://reviewable.io/reviews/intel-isl/open3d/2407#-MIZNxzeiBXMt1ozPAlN-r1-1199:-MIo3qfU1Bu0KYyRKi5n:b-665qkf) ([raw file](https://github.com/intel-isl/open3d/blob/62b95da12842fb6e3c7630b80d40a580ee8525d5/cpp/open3d/geometry/TriangleMesh.cpp#L1199)):*
<details><summary><i>Previously, kongsgard (Sondre Kongsgård) wrote…</i></summary><blockquote>

I'll try to add this after #2408 is closed. I haven't used OpenMP before, but based on your hint I think the following should suffice:
```
double volume = 0;
size_t num_triangles = triangles_.size();
#pragma omp parallel for shared(volume, num_triangles) reduction(+ : volume)
```
</blockquote></details>

The build now seems to be successful on all platforms. `int64_t` was the only option that worked, as unsigned integers aren't supported on MSVC compilers (as you mentioned) and `int` might lose some precision.
___
*[cpp/open3d/geometry/TriangleMesh.cpp, line 1204 at r3](https://reviewable.io/reviews/intel-isl/open3d/2407#-MIdsGEE1uPgBd9MLIF9:-MIo3R-J1IW10WJPcidX:b-tn322b) ([raw file](https://github.com/intel-isl/open3d/blob/7dfb92e37e44573df24fd5ba23fee72c72b0214c/cpp/open3d/geometry/TriangleMesh.cpp#L1204)):*
<details><summary><i>Previously, griegler (Gernot) wrote…</i></summary><blockquote>

What I get from the paper `IsWatertight` is not sufficient, I think you should also check `IsOrientable`.
</blockquote></details>

Added a guard for `IsOrientable` too.

**Review by yxlao — APPROVED** [review]

**Review by griegler — APPROVED** [review]
> Reviewed 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2407)* status: all files reviewed, 3 unresolved discussions (waiting on @yxlao)

**Pull request merged**